Docs: Add "How to extend workflows" section (#4562)
aiidateam/aiida-core: docs/source/howto/workflows.rst

How to run multi-step workflows
*******************************

.. _how-to:workflows:run:

Launching a predefined workflow
===============================

The first step to launching a predefined workflow is loading the work function or work chain class that defines the workflow you want to run.
The recommended method for loading a workflow is using the ``WorkflowFactory``, for example:

.. code-block:: python

    from aiida.plugins import WorkflowFactory
    add_and_multiply = WorkflowFactory('arithmetic.add_multiply')

>     MultiplyAddWorkChain = WorkflowFactory('arithmetic.multiply_add')

Author: Sure, if you think it helps to mention this. Maybe I should add it as a `.. note::` directive?

>     MultiplyAddWorkChain = WorkflowFactory('arithmetic.multiply_add')

Reviewer: Yes, up to you. I just think that since this is a recent change it makes sense to consider people who might be using older version.

So, it is advisable to *submit* more complex or longer work chains to the daemon:

.. code-block:: python

    from aiida.engine import submit
    from aiida.plugins import WorkflowFactory, DataFactory
    Int = DataFactory('int')
    MultiplyAddWorkChain = WorkflowFactory('arithmetic.multiply_add')

    add_code = load_code(label='add')
    inputs = {'x': Int(1), 'y': Int(2), 'z': Int(3), 'code': add_code}

    workchain_node = submit(MultiplyAddWorkChain, **inputs)

Note that when using ``submit`` the work chain is not run in the local interpreter but is sent off to the daemon and you get back control instantly.
This allows you to submit multiple work chains at the same time and the daemon will start working on them in parallel.
Once the ``submit`` call returns, you will not get the result as with ``run``, but you will get the **node** that represents the work chain.
Submitting a work chain instead of directly running it not only makes it easier to execute multiple work chains in parallel, but also ensures that the progress of a workchain is not lost when you restart your computer.

.. important::

>     In contrast to work chains, work *functions* cannot be submitted to the daemon, and hence can only be *run*.

Author: Ah, good to know! 😅  Should we keep on mentioning the changes in `v1.5.0` though? Maybe we should just write the documentation for the current version, and not put in these notes about what changed. I.e. in this case just remove the whole `.. important::` directive.

>     In contrast to work chains, work *functions* cannot be submitted to the daemon, and hence can only be *run*.

Reviewer: Yeah, you can choose one place to make the disclaimer and add both things there (plugin + submit). BTW if you are not familiar with this, maybe we should check with @sphuber about this. I'm pretty sure they are now submitable and were not previously, but I'm less sure if not being submitable meant that previously they didn't have entry points or they could also be added to the registry regardless of being submitable or not.

>     In contrast to work chains, work *functions* cannot be submitted to the daemon, and hence can only be *run*.

Reviewer 2: having entry points are not what makes something submittable. The reason that calculation and work functions were not submittable was that they weren't treated as a normal process, back in the day. Now, they are actually identical from the engine's point of view. When launched the function is dynamically turned into an actual `Process` class and so functions exactly like a `WorkChain`. Not that usual caveats apply, the function needs to be importable by the daemon, either through entry points or by making sure the module is in the Python path.

Another note, although process functions technically became submittable only in v1.5, it was technically possible for a lot longer, maybe even as soon as v1.0. In v1.5 we merely removed an explicit check that prevented them from being submitted. Still the result is the same, before v1.5 submitting a process function will raise an error. I think it is fine to keep in the explicit reference of when it became available. People on an older version might be reading the latest version of the docs.

This section explains how to extend workflows by wrapping them around other processes or linking them together.

As an example, let's say you want to extend the ``MultiplyAddWorkChain`` by adding another step of analysis that checks whether the result is an even number or not.
This final step can be written as a simple ``calcfunction``:

.. literalinclude:: include/snippets/extend_workflows.py
    :language: python
    :pyobject: is_even

We could simply write a new workflow based off ``MultiplyAddWorkChain`` that includes an extra step in the outline which runs the ``is_even`` calculation function.
However, this would lead to a lot of code duplication, and longer workflows consisting of multiple work chains would become very cumbersome to deal with (see the dropdown panel below).

.. dropdown:: ``BadMultiplyAddIsEvenWorkChain``

    .. literalinclude:: include/snippets/extend_workflows.py
        :language: python
        :pyobject: BadMultiplyAddIsEvenWorkChain
.. note::

    We've removed the ``result`` step from the outline, as well as the ``result`` output.
    For this work chain, we're assuming that for now we are only interested in whether or not the result is even.

> We can avoid some code duplication by simply submitting the ``MultiplyAddWorkChain`` within one of the steps of our new work chain.
> In order to do this, we replace the ``multiply``, ``add`` and ``validate_result`` steps in the outline with one ``multiply_add`` step, in which we ``submit`` the ``MultiplyAddWorkChain`` and add the returned ``WorkChainNode`` to the context:

Reviewer: If you are hiding code using the dropdown option, then I would say that the rest of the external explanation should not depend on this content. So, concretely, I would put the comments of the note above inside the dropdown and maybe describe the "good" workchain as a new thing instead of a modification of the "bad" one:

```suggestion
We can avoid some code duplication by simply submitting the ``MultiplyAddWorkChain`` within one of the steps of a new work chain which would then call `is_even` in a second step.
```

.. literalinclude:: include/snippets/extend_workflows.py
    :language: python
    :pyobject: BetterMultiplyAddIsEvenWorkChain

This already simplifies the extended work chain, and avoids duplicating the steps of the ``MultiplyAddWorkChain`` in the outline.
However, we still had copy all of the input definitions of the ``MultiplyAddWorkChain``, and manually extract them from the inputs before passing them to the ``self.submit`` method.
Fortunately, there is a better way of *exposing* the inputs and outputs of subprocesses of the work chain.

Exposing inputs and outputs
---------------------------

> Any work chain *needs* to *expose* the inputs of the subprocess it wraps, and perhaps *wants* to do the same for the outputs it produces, although the latter is not necessary.

Author: Fair point! After all, `is_even` is a subprocess of the work chain, and we do not expose its inputs. Let me see if I can rewrite this so it's still clear which inputs need exposing.

> Any work chain *needs* to *expose* the inputs of the subprocess it wraps, and perhaps *wants* to do the same for the outputs it produces, although the latter is not necessary.

Author: Not sure how to explain this without going into an excessive tangent discussion. Will just make the sentence less _needy_.

> Any work chain *needs* to *expose* the inputs of the subprocess it wraps, and perhaps *wants* to do the same for the outputs it produces, although the latter is not necessary.

Reviewer: mmmm, maybe something like:

```
 In many cases it is convenient for workchains to expose the inputs of the subprocess it wraps so that users can cutomize these runs directly, as well as exposing some of the outputs produced as the final results of the whole procedure.
 ```

.. code-block:: console

    $ verdi process status 164
    MultiplyAddIsEvenWorkChain<164> Finished [0] [1:is_even]
        ├── MultiplyAddWorkChain<165> Finished [0] [3:result]
        │   ├── multiply<166> Finished [0]
        │   └── ArithmeticAddCalculation<168> Finished [0]
        └── is_even<172> Finished [0]

Note that this command also recursively shows the processes called by the subprocesses of the ``MultiplyAddIsEvenWorkChain`` work chain.

As mentioned earlier, you can also expose the outputs of the ``MultiplyAddWorkChain`` using the :meth:`~plumpy.ProcessSpec.expose_outputs` method.
Say we want to add the ``result`` of the ``MultiplyAddWorkChain`` as one of the outputs of the extended work chain:

.. code-block:: python

    @classmethod
    def define(cls, spec):
        """Specify inputs and outputs."""
        super().define(spec)
        spec.expose_inputs(MultiplyAddWorkChain, namespace='multiply_add')
        spec.outline(
            cls.multiply_add,
            cls.is_even,
        )
        spec.expose_outputs(MultiplyAddWorkChain)
        spec.output('is_even', valid_type=Bool)

> Since there is not one output port that is shared by all process classes, it is less critical to use the ``namespace`` argument when exposing outputs.

Author: Hmm, I'm not sure if we should recommend it, but an extra sentence as a warning is definitely welcome.

> Since there is not one output port that is shared by all process classes, it is less critical to use the ``namespace`` argument when exposing outputs.

Reviewer: Yeah, now that I think about it maybe you are right, you would tipically handpick the results of the workchain rather than just exposing them directly. Even in complex workchains you won't want to just vomit every output produced in the middle, they can anyways be accessed by exploring the graph (unlike inputs which you need to expose so that users can customize the runs).

Feel free to disregard this, or maybe briefly express this idea I rambled in the previous paragraph.
